test/integration/tcp_proxy_odcds_integration_test.cc

```diff
@@ -0,0 +1,259 @@
+  }
+
+  // The delta CDS stream is established.
+  auto result = fake_upstreams_.front()->waitForHttpConnection(*dispatcher_, xds_connection_);
+  RELEASE_ASSERT(result, result.message());
+  result = xds_connection_->waitForNewStream(*dispatcher_, odcds_stream_);
+  RELEASE_ASSERT(result, result.message());
+  odcds_stream_->startGrpcStream();
+
+  // Verify the delta CDS request and respond without providing the cluster.
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {"new_cluster"}, {},
+                                           odcds_stream_));
+  sendDeltaDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+      Config::TypeUrl::get().Cluster, {new_cluster_}, {}, "1", odcds_stream_);
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {}, {}, odcds_stream_));
+
+  // This upstream is listening on the endpoint of ``new_cluster``.
+  for (auto n = expected_upstream_connections_; n != 0; n--) {
+    fake_upstream_connections_.push_back(nullptr);
+    auto& fake_upstream_connection = fake_upstream_connections_.back();
+    ASSERT_TRUE(fake_upstreams_.back()->waitForRawConnection(fake_upstream_connection));
+    ASSERT_TRUE(fake_upstream_connection->write("hello"));
+  }
+
+  for (auto& tcp_client : tcp_clients_) {
+    tcp_client->waitForData("hello");
+    ASSERT_TRUE(tcp_client->write("world"));
+  }
+
+  for (auto& fake_upstream_connection : fake_upstream_connections_) {
+    ASSERT_TRUE(fake_upstream_connection->waitForData(5));
+    ASSERT_TRUE(fake_upstream_connection->close());
+    ASSERT_TRUE(fake_upstream_connection->waitForDisconnect());
+  }
+
+  for (auto& tcp_client : tcp_clients_) {
+    tcp_client->waitForHalfClose();
+    tcp_client->close();
+  }
+}
+
+TEST_P(TcpProxyOdcdsIntegrationTest, ShutdownConnectionOnTimeout) {
+  initialize();
+
+  // Establish a tcp request to the Envoy.
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("tcp_proxy"));
+
+  // The delta CDS stream is established.
+  auto result = fake_upstreams_.front()->waitForHttpConnection(*dispatcher_, xds_connection_);
+  RELEASE_ASSERT(result, result.message());
+  result = xds_connection_->waitForNewStream(*dispatcher_, odcds_stream_);
+  RELEASE_ASSERT(result, result.message());
+  odcds_stream_->startGrpcStream();
+
+  // Verify the delta CDS request and respond without providing the cluster.
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {"new_cluster"}, {},
+                                           odcds_stream_));
+  sendDeltaDiscoveryResponse<envoy::config::cluster::v3::Cluster>(Config::TypeUrl::get().Cluster,
+                                                                  {}, {}, "1", odcds_stream_);
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {}, {}, odcds_stream_));
+
+  tcp_client->waitForHalfClose();
+  tcp_client->close();
+}
+
+TEST_P(TcpProxyOdcdsIntegrationTest, ShutdownConnectionOnClusterMissing) {
```

Comment:
do we have any stats on what went wrong folks could use for debug?  if not consider adding some?

Reply:
adding...

Reply:
add 3 counters and check in integration tests
on_demand_cluster_attempt
on_demand_cluster_missing
on_demand_cluster_timeout